### PR TITLE
[CoSim] variable redistributen operations

### PR DESCRIPTION
**Pull request by philbucher**
For mapping of fields: conversion from concentrated nodal quantities to fields and vise-versa

@shahrokh-shayegan I have not tested it yet, we should speak in person

### CODE REVIEW

File: applications/CoSimulationApplication/python_scripts/coupling_operations/convert_distributed_values_to_point.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    @classmethod
+    def _GetDefaultSettings(cls):
+        this_defaults = KM.Parameters("""{
+            "solver"                       : "UNSPECIFIED",
+            "data_name"                    : "UNSPECIFIED",
+            "intermediate_variable_scalar" : "NODAL_PAUX",
+            "intermediate_variable_vector" : "NODAL_VAUX"
+        }""")
+        this_defaults.AddMissingParameters(super(DistributePointValuesOperation, cls)._GetDefaultSettings())
```
Comment (shahrokh-shayegan) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/convert_distributed_values_to_point.py]:
```suggestion
        this_defaults.AddMissingParameters(super(ConvertDistributedValuesToPoint, cls)._GetDefaultSettings())
```

File: applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py [hunk 2]
```diff
@@ -1,38 +1,65 @@
...
 
-    def FinalizeCouplingIteration(self):
-        pass
+    def Execute(self):
+        KM.VariableRedistributionUtility.DistributePointValues(
+            self.interface_data.GetModelPart(),
+            self.intermediate_variable,
+            self.interface_data.variable,
+            self.redistribution_tolerance,
+            self.redistribution_iterations)
```
Comment (shahrokh-shayegan) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py]: I think the order of the 2nd and the 3rd arguments of this function should be changed as well.
for example in a Dirichlet-Neumann coupling, this function would be used on the fluid's FSI interface to convert REACTION as interface_data.variable to "distributed tractions" in an intermediate auxiliary variable.

File: applications/CoSimulationApplication/python_scripts/coupling_operations/convert_distributed_values_to_point.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+from __future__ import print_function, absolute_import, division  # makes these scripts backward compatible with python 2.6 and 2.7
+
+# Importing the Kratos Library
+import KratosMultiphysics as KM
+
+# Importing the base class
+from KratosMultiphysics.CoSimulationApplication.base_classes.co_simulation_coupling_operation import CoSimulationCouplingOperation
+
```
Comment (shahrokh-shayegan) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/convert_distributed_values_to_point.py:6]:
```suggestion
# CoSimulation imports
import KratosMultiphysics.CoSimulationApplication.co_simulation_tools as cs_tools

def Create(settings, solver_wrappers):
    cs_tools.SettingsTypeCheck(settings)
    return ConvertDistributedValuesToPoint(settings, solver_wrappers)
```
missing Create function

File: applications/CoSimulationApplication/python_scripts/coupling_operations/convert_distributed_values_to_point.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+class ConvertDistributedValuesToPoint(CoSimulationCouplingOperation):
+
+    def __init__(self, settings, solver_wrappers):
+        super(ComputeNormalsOperation, self).__init__(settings)
```
Comment (shahrokh-shayegan) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/convert_distributed_values_to_point.py]:
```suggestion
        super(ConvertDistributedValuesToPoint, self).__init__(settings)
```

File: applications/CoSimulationApplication/python_scripts/coupling_operations/convert_distributed_values_to_point.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        solver_name = self.settings["solver"].GetString()
+        data_name = self.settings["data_name"].GetString()
+        self.interface_data = solver_wrappers[solver_name].GetInterfaceData(data_name)
+
+        if self.interface_data.is_scalar_variable:
+            self.intermediate_variable = KM.KratosGlobals.GetVariable(self.settings["intermediate_variable_scalar"].GetString())
+        else:
+            self.intermediate_variable = KM.KratosGlobals.GetVariable(self.settings["intermediate_variable_scalar"].GetString())
```
Comment (shahrokh-shayegan) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/convert_distributed_values_to_point.py]:
```suggestion
            self.intermediate_variable = KM.KratosGlobals.GetVariable(self.settings["intermediate_variable_vector"].GetString())
```

File: applications/CoSimulationApplication/python_scripts/coupling_operations/convert_distributed_values_to_point.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    def Execute(self):
+        KM.VariableRedistributionUtility.ConvertDistributedValuesToPoint(
+            self.interface_data.GetModelPart(),
+            self.interface_data.variable,
+            self.intermediate_variable)
```
Comment (shahrokh-shayegan) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/convert_distributed_values_to_point.py]: I think the order of the last two arguments of this function should be changed.
for example in a Dirichlet-Neumann coupling, this function would be used on the structure's FSI interface to convert distributed tractions from an intermediate auxiliary variable to POINT_LOAD.

File: applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py [hunk 7]
```diff
@@ -1,38 +1,65 @@
 from __future__ import print_function, absolute_import, division  # makes these scripts backward compatible with python 2.6 and 2.7
 
+# Importing the Kratos Library
+import KratosMultiphysics as KM
+
 # Importing the base class
 from KratosMultiphysics.CoSimulationApplication.base_classes.co_simulation_coupling_operation import CoSimulationCouplingOperation
```
Comment (shahrokh-shayegan) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py:6]:
```suggestion
# CoSimulation imports
import KratosMultiphysics.CoSimulationApplication.co_simulation_tools as cs_tools

def Create(settings, solver_wrappers):
    cs_tools.SettingsTypeCheck(settings)
    return DistributePointValuesOperation(settings, solver_wrappers)
```
missing Create function

File: applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py [hunk 8]
```diff
@@ -1,38 +1,65 @@
...
-        pass
-
-
-    def InitializeSolutionStep(self):
-        pass
 
-    def FinalizeSolutionStep(self):
-        pass
+    def __init__(self, settings, solver_wrappers):
+        super(ComputeNormalsOperation, self).__init__(settings)
```
Comment (shahrokh-shayegan) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py]:
```suggestion
       super(DistributePointValuesOperation, self).__init__(settings)
```

File: applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py [hunk 9]
```diff
@@ -1,38 +1,65 @@
...
+        solver_name = self.settings["solver"].GetString()
+        data_name = self.settings["data_name"].GetString()
+        self.interface_data = solver_wrappers[solver_name].GetInterfaceData(data_name)
 
+        self.redistribution_iterations = self.settings["redistribution_tolerance"].GetDouble()
```
Comment (shahrokh-shayegan) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py]:
```suggestion
       self.redistribution_iterations = self.settings["redistribution_iterations"].GetInt()
```

File: applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py [hunk 10]
```diff
@@ -1,38 +1,65 @@
...
+        self.redistribution_tolerance  = self.settings["redistribution_iterations"].GetInt()
```
Comment (shahrokh-shayegan) [inline applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py]:
```suggestion
        self.redistribution_tolerance  = self.settings["redistribution_tolerance"].GetDouble()
```

Comment (shahrokh-shayegan) [reply applications/CoSimulationApplication/python_scripts/coupling_operations/distribute_point_values.py]: for testing the implementation, I added the required variables for the fluid model part (MAPPER_VECTOR_PROJECTION_RHS, NODAL_VAUX) inside the fluid solver, which is obviously not a good idea. So maybe one should think of a good place to add the extra variables to the fluid model part.

**Comment by shahrokh-shayegan** [comment]
for testing the implementation, I added the required variables for the fluid model part (MAPPER_VECTOR_PROJECTION_RHS, NODAL_VAUX) inside the fluid solver, which is obviously not a good idea. So maybe one should think of a good place to add the extra variables to the fluid model part.

**Comment by philbucher** [comment]
ok this is finally ready :tada:

**Review by mpentek — COMMENTED** [review]
Thanks

**Review by mpentek — APPROVED** [review]
Thanks

**Pull request merged**